### PR TITLE
Fix issue when compiling with HIP

### PR DESCRIPTION
**Pull request by Rombur**
With this PR I can compile and run deal.II on Frontier.

### CODE REVIEW

File: include/deal.II/base/numbers.h [hunk 1]
```diff
@@ -41,7 +41,7 @@
 // This is problematic when calling non-constexpr functions in constexpr
 // functions. Hence, we need a way to annotate functions explicitly as
 // host-only.
-#if defined(__clang__) && defined(__CUDA__)
+#if defined(__clang__) && defined(__CUDA__) || defined(KOKKOS_ENABLE_HIP)
```
Comment (masterleinad) [inline include/deal.II/base/numbers.h:44]: I always forget if `&&` has precedence over `||`.
```suggestion
#if (defined(__clang__) && defined(__CUDA__)) || defined(KOKKOS_ENABLE_HIP)
```
Also, the above command also doesn't match anymore. Maybe, say something like
```
// clang++ assumes that all constexpr functions are __host__ __device__ when `Kokkos` was configured with  `HIP` or `CUDA` support. [...]
```

Comment (Rombur) [reply include/deal.II/base/numbers.h:44]: > I always forget if && has precedence over ||

It doesn't matter since all the HIP compiler are clang based

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/base/numbers.h [hunk 2]
```diff
@@ -37,11 +37,10 @@
 #define DEAL_II_HOST_DEVICE_ALWAYS_INLINE KOKKOS_FORCEINLINE_FUNCTION
 
 // clang++ assumes that all constexpr functions are __host__ __device__ when
-// compiling CUDA code, i.e, when Kokkos was configured with CUDA support.
-// This is problematic when calling non-constexpr functions in constexpr
-// functions. Hence, we need a way to annotate functions explicitly as
-// host-only.
-#if defined(__clang__) && defined(__CUDA__)
+// when Kokkos was configured with CUDA or HIP support. This is problematic
```
Comment (marcfehling) [inline include/deal.II/base/numbers.h:40]:
```suggestion
// Kokkos was configured with CUDA or HIP support. This is problematic
```
Just minor documentation comment: duplicated from pervious line.

**Review by marcfehling — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Pull request merged**